Network/Converter/PacketConverter.cs

```diff
@@ -27,321 +28,680 @@
+            }
+
+            // we have a non-primitive type
+            if (!PacketConverterHelper.TypeIsPrimitive(propertyType))
+            {
+                NetworkObjectState networkObjectState =
+                    (NetworkObjectState)binaryReader.ReadByte();
+
+                if (networkObjectState == NetworkObjectState.NotNull)
+                {
+                    // this will recursively deserialise all the properties
+                    // that are made up of primitives (even complex types)
+                    return DeserialiseObjectFromReader(
+                        PacketConverterHelper.InstantiateObject(propertyType), binaryReader);
+                }
+
+                // if it is null we just return null
+                return null;
+            }
+
+            // we have a primitive type
+            return ReadPrimitiveFromStream(propertyInfo, binaryReader);
         }
 
         /// <summary>
-        /// Reads all the properties from an object and calls the <see cref="ReadObjectFromStream"/> method.
+        /// Deserialises the given <see cref="Array"/> from the given
+        /// <see cref="BinaryReader"/>s underlying <see cref="MemoryStream"/>.
+        /// Uses <see cref="DeserialiseObjectFromReader(object,BinaryReader)"/>
+        /// to serialise each of the <see cref="Array"/>s elements to the stream.
         /// </summary>
-        /// <param name="obj">The object.</param>
-        /// <param name="binaryReader">The binary reader.</param>
-        /// <returns>System.Object.</returns>
-        private object ReadObjectFromStream(object obj, BinaryReader binaryReader)
+        /// <param name="obj">
+        /// The <see cref="Array"/> to deserialise from the given
+        /// <see cref="BinaryReader"/>s underlying <see cref="MemoryStream"/>.
+        /// </param>
+        /// The <see cref="PropertyInfo"/> holding the <see cref="Array"/>.
+        /// <param name="propertyInfo">
+        /// </param>
+        /// <param name="binaryReader">
+        /// The <see cref="BinaryReader"/> from whose underlying <see cref="MemoryStream"/>
+        /// to deserialise the given <see cref="PropertyInfo"/>.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if the <see cref="Array"/>s elements do not have a type.
+        /// </exception>
+        private Array ReadArrayFromStream(
+            object obj, PropertyInfo propertyInfo, BinaryReader binaryReader)
         {
-            GetPacketProperties(obj).ToList().ForEach(p => p.SetValue(obj, ReadObjectFromStream(obj, p, binaryReader)));
-            return obj; //All properties set with in place.
+            int arraySize = binaryReader.ReadInt32();
+
+            Type elementType = propertyInfo.PropertyType.GetElementType();
+            Array array = Array.CreateInstance(elementType, arraySize);
+
+            for (int i = 0; i < arraySize; ++i)
```

Comment:
Did look it up. ++i and i++ in for equals the same in C#. However, use i++

Reply:
++i might be more efficient, will get back to you on this.